### PR TITLE
Add System Test for HeliumAnalyserEfficiency

### LINKED ISSUE #37225: Add a system test for HeliumAnalyserEfficiency
**Describe the outcome that is desired.**

This needs a system test that runs through the whole algorithm once we're certain of what the inputs and outputs should be.

**Data**

Any run number between `ZOOM38249` and `ZOOM38333` can be passed as the input workspace group. All the even run numbers starting (and including) `ZOOM38253` are unpolarised, the rest are polarised. We should pass `00,10,11,01` to the SpinStates parameter of the algorithm for these datasets.

We have confirmed that these runs are suitable to be made public.

Some pre-processing of the raw data is needed before the workspace group will be ready to pass to the algorithm:

- Convert to wavelength.
- Extract the spectrum corresponding to monitor 3 into it's own workspace group (this is the normalising monitor). This should be spectrum 3.
- Extract the spectrum corresponding to monitor 4 into it's own workspace group (this is the monitor between the analyser and main detector that is used for the transmission measurement). This should be spectrum 4.
- Normalise the monitor 4 workspace group by the monitor 3 workspace group (i.e. mon4WsGrp/mon3WsGrp).  The normalised workspace group is now ready to use in the algorithm.

Examples of the required pre-processing can be seen in Diego's example script.

**Additional context**

See #36999

### PR DESCRIPTION
**Pull request by cailafinn**
#### Summary of work
Adds a system test to confirm the accuracy of the HeliumAnalyserEfficiency algorithm as part of the Polarised SANS epic. 

Fixes #37225 

#### Further detail of work
- Much of the functionality is duplicated between system tests, so an abstract base class has been added to handle much of the pre-processing that is required for these system tests. Future tests can simply inherit from the base class and override the abstract methods. 

### To test:
1. Ensure both new tests can be run from the `systemtest` call in the repo
2. Ensure the refactored DepolarizedAnalyser test also runs after the refactor
3. Check the CI ran the 2 new tests and the Depolarized one. 

*This does not require release notes* because **it only adds system tests, which are dev facing. **

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Comment by sf1919** [comment]
Can we get this run on `build_packages_from_branch` please? Systems tests only get checked on Linux through automated PR testing. I'd like to make sure this will also run on osx and windows too before it gets merged in. This is expecially given all the problems with the nightly pipelines at the moment.

File: Testing/SystemTests/tests/framework/PolarizationCorrections/HeliumAnalyserEfficiencyTest.py
Comment (rbauststfc) [inline Testing/SystemTests/tests/framework/PolarizationCorrections/HeliumAnalyserEfficiencyTest.py:44]: I'm wondering if it would be a bit clearer for the `HeliumAnalyserEfficiencyTest` class not to have a dual function as a base class and a test, and for it instead to just be a base that is inherited by two test classes, `HeliumAnalyserEfficiencyUnpolarisedTest` and `HeliumAnalyserEfficiencyPolarisedTest`?

File: Testing/SystemTests/tests/framework/PolarizationCorrections/PolarizationCorrectionsBase.py
Comment (rbauststfc) [inline Testing/SystemTests/tests/framework/PolarizationCorrections/PolarizationCorrectionsBase.py:23]: This set of pre-processing steps is very specific to the polarised SANS work. Would it be clearer to name this class `SANSPolarizationCorrectionsBase`?

**Review by rbauststfc — COMMENTED** [review]
Really great to see this being set up in a way that will allow us to re-use the logic across all of the SANS system tests, with an easy to maintain and extensible design.

I've made just a few suggestions below that you might like to consider. I also agree with Sarah's comment that we should run a build from branch to check that the tests pass across all operating systems.

File: Testing/SystemTests/tests/framework/PolarizationCorrections/HeliumAnalyserEfficiencyTest.py
Comment (cailafinn) [reply Testing/SystemTests/tests/framework/PolarizationCorrections/HeliumAnalyserEfficiencyTest.py:19]: I'd missed that this could be done automatically. Another system test I'd looked at had re-implemented it and so I thought I needed to so too. Thanks for pointing this out!

File: Testing/SystemTests/tests/framework/PolarizationCorrections/PolarizationCorrectionsBase.py
Comment (cailafinn) [reply Testing/SystemTests/tests/framework/PolarizationCorrections/PolarizationCorrectionsBase.py:43]: I'd missed that this could be done automatically. Another system test I'd looked at had re-implemented it and so I thought I needed to do so too. Thanks for pointing this out!

**Comment by cailafinn** [comment]
Good call on doing a build from branch. Currently running here: https://builds.mantidproject.org/job/build_packages_from_branch/

File: Testing/SystemTests/tests/framework/PolarizationCorrections/HeliumAnalyserEfficiencyTest.py
Comment (rbauststfc) [inline Testing/SystemTests/tests/framework/PolarizationCorrections/HeliumAnalyserEfficiencyTest.py:9]: In several of these files we seem to import all of the `simpleapi`. I realise this may have just been copied across from other tests, but would it be worth just importing the algorithms that we're using?

**Review by rbauststfc — COMMENTED** [review]
This is all looking really good now, thanks! The build from branch looks like it's failed but on something unrelated, the specific tests added in this PR all passed OK. I've also run the tests on my local Windows machine and they passed without an issue.

I know this PR is holding up other work and there's nothing left that needs to be addressed from my perspective. There are a couple of very minor suggestions I noticed on my second look through that could easily be amended in the work that follows, if you felt it was worth it, so I've mentioned those below.

If you want to go ahead with this as it is now then just let me know and I'll be very happy to approve.

**Comment by sf1919** [comment]
I can confirm the test failures in the build packages from branch are unrelated and were fixed in a PR that went in yesterday (therefore after the creation of this branch). I'm re-running the tests on the PR as they got stuck on a dodgy runner but once they all pass if @rbauststfc is happy to approve then thats fine.

**Review by rbauststfc — APPROVED** [review]
All the code changes are looking really good, the tests are passing on this PR and we're happy with the results from the build from branch, so this is good to go in. 👍

**Pull request merged**
